File: src/main/java/xf/xflp/base/container/ContainerBaseData.java
package xf.xflp.base.container;

import util.collection.LPListMap;
import xf.xflp.base.item.Position;

import java.util.Map;

/**
 * Copyright (c) 2012-2026 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @author hschneid
 */
public interface ContainerBaseData {

    LPListMap<Integer, Integer> getXMap();
    LPListMap<Integer, Integer> getYMap();
    LPListMap<Integer, Integer> getZMap();

    ZItemGraph getZGraph();

    Map<Integer, Float> getBearingCapacities();

    /**
     * Returns the precomputed minimum immersive depth for each active position.
     * Key is the Position, value is the minimum immersive depth of all items
     * whose top face is at pos.z() and whose footprint contains the point (pos.x(), pos.y()).
     */
    int getImmersiveDepthAtPosition(Position position);

    float getCenterOfGravityForY();

    /** Returns the maximum yl (y + length) of all items currently in the container. */
    int getMaxYl();
}

File: src/xflp/base/container/container_base_data.py
"""
Container base data interface protocol.

Copyright (c) 2012-2025 Holger Schneider
Licensed under the MIT License
"""

from typing import Protocol, Dict, TYPE_CHECKING

if TYPE_CHECKING:
    from ...util.collection.lp_list_map import LPListMap
    from .z_item_graph import ZItemGraph


class ContainerBaseData(Protocol):
    """
    Protocol for accessing internal container data structures.
    
    Provides access to coordinate maps, item graph, and bearing capacities.
    """
    
    def get_x_map(self) -> 'LPListMap[int, int]':
        """Get the X-coordinate to item index map."""
        ...
    
    def get_y_map(self) -> 'LPListMap[int, int]':
        """Get the Y-coordinate to item index map."""
        ...
    
    def get_z_map(self) -> 'LPListMap[int, int]':
        """Get the Z-coordinate to item index map."""
        ...
    
    def get_z_graph(self) -> 'ZItemGraph':
        """Get the Z-axis item relationship graph."""
        ...
    
    def get_bearing_capacities(self) -> Dict[int, float]:
        """Get the bearing capacity map for items."""
        ...
    
    def get_center_of_gravity_for_y(self) -> float:
        """Get the center of gravity along the Y axis."""
        ...
